include/cuco/detail/probe_sequences.cuh

```diff
@@ -0,0 +1,165 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda/std/atomic>
+
+#include <cuco/detail/hash_functions.cuh>
+#include <cuco/detail/pair.cuh>
+
+namespace cuco {
+
+template <typename Key,
+          typename Value,
+          uint32_t CGSize          = 8,
+          cuda::thread_scope Scope = cuda::thread_scope_device>
+class probe_sequence_base {
+ protected:
+  using value_type         = cuco::pair_type<Key, Value>;
+  using key_type           = Key;
+  using mapped_type        = Value;
+  using atomic_key_type    = cuda::atomic<key_type, Scope>;
+  using atomic_mapped_type = cuda::atomic<mapped_type, Scope>;
+  using pair_atomic_type   = cuco::pair_type<atomic_key_type, atomic_mapped_type>;
+  using iterator           = pair_atomic_type*;
+  using const_iterator     = pair_atomic_type const*;
+
+ public:
+  __host__ __device__ static constexpr uint32_t cg_size() noexcept { return CGSize; }
+
+  __host__ __device__ static constexpr uint32_t vector_width() noexcept { return 2u; }
+
+  __host__ __device__ static constexpr bool uses_vector_load() noexcept
+  {
+    return cuco::detail::is_packable<value_type>();
+  }
+
+  __host__ __device__ explicit probe_sequence_base(iterator slots, std::size_t capacity)
+    : slots_{slots}, capacity_{capacity}
+  {
+  }
+
+  __host__ __device__ std::size_t get_capacity() const noexcept { return capacity_; }
+
+  __device__ iterator get_slots() noexcept { return slots_; }
+  __device__ const_iterator get_slots() const noexcept { return slots_; }
+
+ protected:
+  iterator slots_;
+  const std::size_t capacity_;
+};  // class probe_sequence_base
+
+template <typename Key,
+          typename Value,
+          uint32_t CGSize          = 8,
+          typename Hash1           = cuco::detail::MurmurHash3_32<Key>,
+          typename Hash2           = cuco::detail::MurmurHash3_32<Key>,
+          cuda::thread_scope Scope = cuda::thread_scope_device>
+class double_hashing : public probe_sequence_base<Key, Value, CGSize, Scope> {
+ public:
+  using iterator = typename probe_sequence_base<Key, Value, CGSize, Scope>::iterator;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::capacity_;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::slots_;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::cg_size;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::vector_width;
+  using probe_sequence_base<Key, Value, CGSize, Scope>::uses_vector_load;
+
+  __host__ __device__ explicit double_hashing(iterator slots, std::size_t capacity) noexcept
```

Comment:
Shouldn't the constructor accept `Hash1` and `Hash2` objects instead of requiring all hash functions be default constructible? I'm not sure how this could be working with cuDF's join where the hash function object is stateful.

Reply:
Right. I didn't notice this issue since the new hash join doesn't use cuDF hash function object to build multimap.

Reply:
Oh, right, because we compute the hash ahead of time with a transform iterator.